src/core/records-grant-authorization.ts

```diff
@@ -16,6 +14,7 @@ export class RecordsGrantAuthorization {
   public static async authorizeRecordsGrant(
     tenant: string,
     incomingMessage: RecordsRead | RecordsWrite | RecordsDelete,
+    recordsWrite: RecordsWrite,
```

Comment:
Would be nice to add a `@param` above, like what you've done in line 23, as it is unclear at a glance. Though I personally would probably prefer to structure the code slightly differently, with at least incoming `RecordsWrite` being on its own, even though there would be more code duplication.

Reply:
I decided to split this into two separate methods, `authorizeRead` and `authorizeWrite`, which makes the interface a bit clearer. There is a small amount of code duplication but I was able to DRY it up in private methods.